sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java

```diff
@@ -83,6 +89,284 @@ protected void afterTest() {
+                messageLockToken.set(MessageLockToken.fromString(received.getLockToken()));
+                return messageContext;
+            }))
+            .assertNext(receivedMessage -> assertMessageEquals(receivedMessage, messageId, isSessionEnabled))
+            .verifyComplete();
+
+        // complete the message using lock token only using a receiver which represent a different machine
+        StepVerifier.create(receiverNonConnectionSharing.complete(messageLockToken.get(), transaction.get()))
+            .verifyComplete();
+
+        // rollback the transaction.
+        StepVerifier.create(receiverNonConnectionSharing.rollbackTransaction(transaction.get()))
+            .verifyComplete();
+
+        // read the message back, since it was rolled-back previously.
+        final ServiceBusReceivedMessageContext receivedContext = receiveAndDeleteReceiver.receive().next().block(TIMEOUT);
+        assertMessageEquals(receivedContext, messageId, isSessionEnabled);
+
+        messagesPending.decrementAndGet();
+    }
+
+    /**
+     * Verifies that we can do following using shared connection.
+     * 1. create transaction
+     * 2. send message  with transactionContext
+     * 3. receive and settle with transactionContext.
+     * 4. commit Rollback this transaction.
+     */
+    @MethodSource("messagingEntityTxnWithSessions")
+    @ParameterizedTest
+    void transactionSendReceiveAndSettle(MessagingEntityType entityType, boolean isSessionEnabled,
+         boolean commitTransaction, DispositionStatus dispositionStatus) {
+
+        // Arrange
+        setSenderAndReceiver(entityType, isSessionEnabled, true);
+
+        final String messageId1 = "1";
+        final ServiceBusMessage message1 = getMessage(messageId1, isSessionEnabled);
+        final String messageId2 = "2";
+        final ServiceBusMessage message2 = getMessage(messageId2, isSessionEnabled);
+        final String deadLetterReason = "testing";
+        sendMessage(message1).block(TIMEOUT);
+
+
+        // Assert & Act
+        AtomicReference<ServiceBusTransactionContext> transaction = new AtomicReference<>();
+        StepVerifier.create(receiver.createTransaction())
+            .assertNext(txn -> {
+                transaction.set(txn);
+                assertNotNull(transaction);
+            })
+            .verifyComplete();
+        assertNotNull(transaction.get());
+
+        // Assert & Act
+        StepVerifier.create(sender.send(message2, transaction.get()))
+            .verifyComplete();
+
+        final ServiceBusReceivedMessageContext receivedContext = receiver.receive().next().block(TIMEOUT);
+        assertNotNull(receivedContext);
+
+        final ServiceBusReceivedMessage receivedMessage = receivedContext.getMessage();
+        assertNotNull(receivedMessage);
+        final Mono<Void> operation;
+        if (DispositionStatus.ABANDONED == dispositionStatus && isSessionEnabled) {
```

Comment:
This block of code is hard to follow same with the sync case below. Can we split this into two test cases, one with sessions and one with not sessions.